vm-validator/src/vm_validator.rs

```diff
@@ -156,37 +160,36 @@ impl<S: StateView> ModuleCache for ValidationState<S> {
         };
 
         // Get the state value that exists in the actual state and compute the hash.
-        let state_value = self
+        let state_slot = self
             .state_view
-            .get_state_value(&StateKey::module_id(key))
-            .map_err(|err| module_storage_error!(key.address(), key.name(), err))?
-            .ok_or_else(|| {
+            .get_state_slot(&StateKey::module_id(key))
+            .map_err(|err| module_storage_error!(key.address(), key.name(), err))?;
+        if state_slot.as_state_value_opt().is_none() {
+            return Err(
                 PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
                     .with_message(format!(
                         "Module {}::{} cannot be found in storage, but exists in cache",
                         key.address(),
                         key.name()
                     ))
-                    .finish(Location::Undefined)
-            })?;
-        let hash = sha3_256(state_value.bytes());
-
-        // If hash is the same - we can use the same module from cache. If the hash is different,
-        // then the state contains a newer version of the code. We deserialize the state value and
-        // replace the old cache entry with the new code.
-        // TODO(loader_v2):
-        //   Ideally, commit notification should specify if new modules should be added to the
-        //   cache instead of checking the state view bytes. Revisit.
-        Ok(if module.extension().hash() == &hash {
+                    .finish(Location::Undefined),
+            );
+        }
+        let value_version = state_slot.expect_value_version() as usize;
+        // deserialize only relies on local config, so only need to detect changes on module bytes
+        // if we want to support verified modules, we need
+        // to detect changes on aptos environment too.
+        Ok(if version == value_version {
             Some((module, version))
         } else {
+            let state_value = state_slot.into_state_value_opt().unwrap();
```

Comment:
nit: can't we just have some `unpack` for slot so we can get `Some((version, state_value))`? then we can use it above to return early, and use contents of option here